lib/resolvers/error.js

```diff
@@ -31,14 +32,49 @@ const _cdsToGraphQLError = (context, err) => {
 }
 
 const _ensureError = error => (error instanceof Error ? error : new Error(error))
+const _clone = obj => Object.create(Object.getPrototypeOf(obj), Object.getOwnPropertyDescriptors(obj))
 
 const handleCDSError = (context, error) => {
   error = _ensureError(error)
-  // TODO: choose log level depending on type of error analogous to OData adapter
-  if (LOG_CDS._error) LOG_CDS.error(error)
+  _log(context.req, error)
   return _cdsToGraphQLError(context, error)
 }
 
+const _log = (req, error) => {
+  // do not log standard errors, e.g. TypeError
+  if (error.code === undefined && error.details === undefined) return
+
+  // log errors and warnings only
+  if (LOG_CDS.level <= cds.log.levels.WARN) return
+
+  // Clone of the original error object to prevent mutation and unintended side-effects.
+  // Notice that the cloned error is logged to standard output in its default language,
+  // whether the original error message is locale-dependent as it is usually sent in the
+  // HTTP response to HTTP Clients to be displayed in the user interface.
+  const error2log = _clone(error)
+
+  // interpolate message placeholder to default language
+  error2log.message = getErrorMessage(error)
+  if (error.details) {
+    error2log.details = error.details.map(error => {
+      const clone = _clone(error)
+      clone.message = getErrorMessage(error)
+      return clone
+    })
+  }
+
+  const reqClone = _clone(req)
+  const statusCode = normalizeError(error2log, reqClone).statusCode
+
+  if (statusCode >= 400 && statusCode < 500) {
```

Comment:
I think we can avoid cloning the request and calling `normalizeError` if we instead require and use `isClientError` similar to our logging in okra:

```suggestion
  if (isClientError(error)) {
```

When using that function I don't think we would need `context.req` at all for logging the error with default locale. The only downside is another hard dependency to an internal `@sap/cds` function.

Reply:
I think this wouldn't work as expected. The problem is that the `error` object we receive from the input validation in CAP land doesn't have the `code` resolved to a status code yet; instead, it's a placeholder like `ASSERT_NOT_NULL`, ... .